c_src/exml_event.c

```diff
@@ -215,32 +265,55 @@ static ERL_NIF_TERM parse(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
     enif_get_int(env, argv[2], &is_final);
     enif_inspect_binary(env, argv[1], &stream);
 
-    expat_parser *parser_data = XML_GetUserData((XML_Parser)(*parser));
+    expat_parser *parser_data = XML_GetUserData(*parser);
     parser_data->result = enif_make_list(env, 0);
     parser_data->env = env;
-    XML_SetUserData((XML_Parser)(*parser), parser_data);
+    XML_SetUserData(*parser, parser_data);
 
-    res = XML_Parse((XML_Parser)(*parser), (const char *)stream.data, stream.size, is_final);
-    if(!res)
+    while (1)
```

Comment:
I know this is `C` code and things like `while(1)` are OK in it. Do you think it would be possible and still efficient to rewrite this to avoid `while(1)`?